src/WebJobs.Script.WebHost/Security/KeyManagement/BlobStorageSecretsMigrationRepository.cs

```diff
@@ -0,0 +1,146 @@
+        }
+
+        public override async Task WriteSnapshotAsync(ScriptSecretsType type, string functionName, string secretsContent)
+        {
+            await _blobStorageSecretsRepository.Value.WriteSnapshotAsync(type, functionName, secretsContent);
+        }
+
+        public override async Task PurgeOldSecretsAsync(IList<string> currentFunctions, ILogger logger)
+        {
+            await _blobStorageSecretsRepository.Value.PurgeOldSecretsAsync(currentFunctions, logger);
+        }
+
+        public override async Task<string[]> GetSecretSnapshots(ScriptSecretsType type, string functionName)
+        {
+            return await _blobStorageSecretsRepository.Value.GetSecretSnapshots(type, functionName);
+        }
+
+        public override void Dispose()
+        {
+            _blobStorageSecretsRepository.Value.Dispose();
+        }
+
+        private async Task CopyKeysFromFileSystemToBlobStorage(BlobStorageSecretsRepository blobStorageSecretsRepository)
+        {
+            string migrateSentialPath = Path.Combine(blobStorageSecretsRepository.SecretsSentinelFilePath, "migrate-sentinel.json");
+            if (File.Exists(migrateSentialPath))
+            {
+                // Migration is already done
+                Log("[Migration] Sential file is detected.");
+                return;
+            }
+
+            try
+            {
+                using (var stream = new FileStream(migrateSentialPath, FileMode.CreateNew))
+                using (var writer = new StreamWriter(stream))
+                {
+                    //write file
+                    Log("[Migration] Sential file created.");
+                }
+            }
+            catch (IOException)
+            {
+                Log("[Migration] Sential file is alredy created by another instance.");
+                return;
+            }
+
+            string[] files = Directory.GetFiles(blobStorageSecretsRepository.SecretsSentinelFilePath.Replace("Sentinels", string.Empty));
+            if (await blobStorageSecretsRepository.BlobContainer.ExistsAsync())
+            {
+                BlobResultSegment resultSegment = await blobStorageSecretsRepository.BlobContainer.ListBlobsSegmentedAsync(blobStorageSecretsRepository.SecretsBlobPath + "/", null);
+
+                // Check for conflicts
+                if (resultSegment.Results.ToArray().Length > 0)
+                {
+                    Log("[Migration] Conflict detected. Secrets container is not empty.");
+                    return;
+                }
+            }
+            else
+            {
+                await blobStorageSecretsRepository.BlobContainer.CreateIfNotExistsAsync();
+            }
+
+            if (files.Length > 0)
+            {
+                List<Task> copyTasks = new List<Task>();
+                foreach (string file in files)
+                {
+                    Task copyTask = new Task(() =>
```

Comment:
Shouldn't need to create the task.

Your task to collect (and await on at the end) would be what is returned by `UploadFromFileAsync`